### PR TITLE
Test concurrent calls to CallSite.AddRule() and MoveRule()

### PR DESCRIPTION
**Pull request by cston**
See https://github.com/dotnet/runtime/issues/36983

This PR contains a test only. See https://github.com/dotnet/runtime/pull/39159 for the corresponding fix.

### CODE REVIEW

File: src/libraries/System.Linq.Expressions/tests/Dynamic/CallSiteCachingTests.cs [hunk 1]
```diff
@@ -218,5 +222,63 @@ public void BinderCacheFlushWhenTooBig()
             Assert.NotEqual(rules1, rules3);
 
         }
+
+        [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsThreadingSupported))]
```
Comment (stephentoub) [inline src/libraries/System.Linq.Expressions/tests/Dynamic/CallSiteCachingTests.cs:226]: Outerloop?

Comment (cston) [reply src/libraries/System.Linq.Expressions/tests/Dynamic/CallSiteCachingTests.cs:226]: When are `[OuterLoop]` tests executed?

Comment (stephentoub) [reply src/libraries/System.Linq.Expressions/tests/Dynamic/CallSiteCachingTests.cs:226]: By the developer explicitly locally, on request with an azp command on PRs, and daily-ish on rolling builds.

Comment (cston) [reply src/libraries/System.Linq.Expressions/tests/Dynamic/CallSiteCachingTests.cs:226]: > daily-ish on rolling builds

Thanks. In that case, could we keep the test as-is, in all runs? The test only takes ~0.2 secs to complete.

Comment (stephentoub) [reply src/libraries/System.Linq.Expressions/tests/Dynamic/CallSiteCachingTests.cs:226]: Is that true even on underpowered machines, ARM, etc.? Looks a bit like a stress test. If it'll run that quickly everywhere, then innerloop is fine. Thanks.

**Review by 333fred — APPROVED** [review]

File: src/libraries/System.Linq.Expressions/tests/Dynamic/CallSiteCachingTests.cs (hunk 1 above)
Comment (cston) [reply src/libraries/System.Linq.Expressions/tests/Dynamic/CallSiteCachingTests.cs:226]: > Is that true even on underpowered machines, ARM, etc.?

I haven't found how to determine test execution time on ARM yet. If the execution time is too slow on some platforms, perhaps we could move this test to outer loop, or limit the test to a subset of platforms, in a subsequent PR.

**Pull request merged**